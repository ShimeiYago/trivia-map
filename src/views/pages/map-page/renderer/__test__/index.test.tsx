import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, State, Props } from '..';

let shallowWrapper: ShallowWrapper<Props, State, Renderer>;

const props: Props = {
  isFormEditting: false,
  isMobile: true,
  isFormChangedFromLastSaved: false,
  park: 'S',
  updateFoocusingPark: jest.fn(),
  updateFilteringCategoryId: jest.fn(),
};

const mockAddEventListener = jest
  .spyOn(window, 'addEventListener')
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  .mockImplementation(() => {});

const mockRemoveEventListener = jest
  .spyOn(window, 'removeEventListener')
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  .mockImplementation(() => {});

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
  });

  it('basic', () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('land mode', () => {
    shallowWrapper.setProps({
      park: 'L',
    });
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('pc view', () => {
    shallowWrapper.setProps({
      isMobile: false,
    });
    shallowWrapper.setState({
      openFormModal: true,
    });
    expect(shallowWrapper).toMatchSnapshot();
  });
});

describe('handleClickAddButton', () => {
  it('should set openingModal and edittingPostId states', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    instance['handleClickAddButton']();
    expect(instance.state.openFormModal).toBeTruthy;
    expect(instance.state.readingArticleId).toBe(undefined);
  });
});

describe('componentDidMount', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
    jest.resetAllMocks();
  });

  it('should open form modal', () => {
    shallowWrapper.setProps({
      postIdToEdit: 100,
    });
    const instance = shallowWrapper.instance();

    instance.componentDidMount();
    expect(instance.state.openFormModal).toBeTruthy();
  });
});

describe('componentDidUpdate', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
    jest.resetAllMocks();
  });

  it('should add beforeunload eventListener when form changed', () => {
    shallowWrapper.setProps({
      isFormChangedFromLastSaved: true,
    });
    const instance = shallowWrapper.instance();

    instance.componentDidUpdate({ isFormChangedFromLastSaved: false } as Props, {} as State);
    expect(mockAddEventListener).toHaveBeenCalled();
  });

  it('should remove beforeunload eventListener when form did not changed', () => {
    shallowWrapper.setProps({
      isFormChangedFromLastSaved: false,
    });
    const instance = shallowWrapper.instance();

    instance.componentDidUpdate({ isFormChangedFromLastSaved: true } as Props, {} as State);
    expect(mockRemoveEventListener).toHaveBeenCalled();
  });
});

describe('handleBeforeUnload', () => {
  it('should add beforeunload eventListener when form open', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    const event = {
      preventDefault: jest.fn(),
      returnValue: '',
    } as unknown as BeforeUnloadEvent;

    instance['handleBeforeUnload'](event);

    expect(event.returnValue).toBe('未保存のデータがありますが、本当に閉じてもよろしいですか？');
  });
});

describe('handleClickPostEdit', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
  });

  it('should open dialog when form modal is opening', () => {
    shallowWrapper.setProps({ isFormChangedFromLastSaved: true });
    const instance = shallowWrapper.instance();

    instance['handleClickPostEdit']();
    expect(instance.state.openDoubleEditAlartDialog).toBeTruthy;
  });

  it('should set openingModal and edittingPostId states', () => {
    const instance = shallowWrapper.instance();

    instance.setState({
      readingArticleId: 100,
    });

    instance['handleClickPostEdit']();
    expect(instance.state.openFormModal).toBeTruthy;
    expect(instance.state.edittingPostId).toBe(100);
  });
});

describe('handleClickPostDelete', () => {
  it('should set openDialogToConfirmDeleting state', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    instance['handleClickPostDelete']();
    expect(instance.state.openDialogToConfirmDeleting).toBeTruthy();
  });
});

describe('handleCancelToDeleteMarker', () => {
  it('should set openDialogToConfirmDeleting state', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    instance.setState({
      openDialogToConfirmDeleting: true,
    });

    instance['handleCancelToDeleteMarker']();
    expect(instance.state.openDialogToConfirmDeleting).toBeFalsy();
  });
});

describe('handleOpenEditForm', () => {
  it('should open edit form modal', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    instance['handleOpenEditForm']();
    expect(instance.state.openFormModal).toBeTruthy;
  });
});

describe('handleCloseFormModal', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
  });

  it('should close form modal', () => {
    const instance = shallowWrapper.instance();

    instance.setState({
      openFormModal: true,
    });
    instance['handleCloseFormModal']();
    expect(instance.state.openFormModal).toBeFalsy;
  });
});

describe('handleHideFormModal', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<Renderer {...props} />);
  });

  it('should close form modal', () => {
    const instance = shallowWrapper.instance();

    instance.setState({
      openFormModal: true,
    });
    instance['handleHideFormModal']();
    expect(instance.state.openFormModal).toBeFalsy;
  });
});

describe('startToSelectPosition', () => {
  it('should set states', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();
    instance['startToSelectPosition']();
    expect(instance.state.openFormModal).toBeFalsy;
    expect(instance.state.newMarkerMode).toBe(true);
  });
});

describe('endToSelectPosition', () => {
  it('should set states', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();
    instance['endToSelectPosition']();
    expect(instance.state.openFormModal).toBeTruthy;
    expect(instance.state.newMarkerMode).toBe(false);
  });
});

describe('handleCloseDoubleEditAlartDialog', () => {
  it('should close modal', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    shallowWrapper.setState({ openDoubleEditAlartDialog: true });
    const instance = shallowWrapper.instance();

    instance['handleCloseDoubleEditAlartDialog']();
    expect(instance.state.openDoubleEditAlartDialog).toBeFalsy();
  });
});

describe('handleChangePark', () => {
  it('call updateFocusingPark with "L"', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    const event = {
      target: {
        checked: false,
      },
    };

    instance['handleChangePark'](event as React.ChangeEvent<HTMLInputElement>);
    expect(instance.props.updateFoocusingPark).toBeCalledWith('L');
  });

  it('call updateFocusingPark with "S"', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    const event = {
      target: {
        checked: true,
      },
    };

    instance['handleChangePark'](event as React.ChangeEvent<HTMLInputElement>);
    expect(instance.props.updateFoocusingPark).toBeCalledWith('S');
  });
});

describe('handleClickCategoryButton', () => {
  it('change category state', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    const instance = shallowWrapper.instance();

    instance['handleClickCategoryButton'](1)();
    expect(instance.props.updateFilteringCategoryId).toBeCalledWith(1);
  });

  it('set undefined category state if same button is clicked', () => {
    shallowWrapper = shallow(<Renderer {...props} />);
    shallowWrapper.setProps({
      filteringCategoryId: 1,
    });
    const instance = shallowWrapper.instance();

    instance['handleClickCategoryButton'](1)();
    expect(instance.props.updateFilteringCategoryId).toBeCalledWith(undefined);
  });
});
