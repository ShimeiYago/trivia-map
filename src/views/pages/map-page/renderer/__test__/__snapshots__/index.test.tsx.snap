// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<ForwardRef(Box)
  sx={Object {}}
>
  <GlobalMenu
    topBarPosition="static"
  >
    <ForwardRef(Box)
      sx={
        Object {
          "height": "calc(100vh - 58px)",
          "position": "relative",
        }
      }
    >
      <TriviaMap
        endToSelectPosition={[Function]}
        hiddenMarkerIds={Array []}
        newMarkerMode={false}
        park="S"
        shouldCurrentPositionAsyncWithForm={true}
      />
      <FloatingButton
        color="primary"
        disabled={false}
        icon="add-marker"
        onClick={[Function]}
      />
      <LoadingProgressBar />
      <ForwardRef(Box)
        sx={
          Object {
            "backgroundColor": "white",
            "borderColor": "#9e9e9e",
            "borderRadius": 2,
            "borderStyle": "solid",
            "bottom": "auto",
            "left": "auto",
            "margin": 0,
            "padding": 1,
            "position": "fixed",
            "right": 20,
            "top": 70,
            "zIndex": 1000,
          }
        }
      >
        <ForwardRef(FormControl)
          component="fieldset"
        >
          <ForwardRef(RadioGroup)
            onChange={[Function]}
            value="S"
          >
            <ForwardRef(FormControlLabel)
              control={<ForwardRef(Radio) />}
              label="ランド"
              value="L"
            />
            <ForwardRef(FormControlLabel)
              control={<ForwardRef(Radio) />}
              label="シー"
              value="S"
            />
          </ForwardRef(RadioGroup)>
        </ForwardRef(FormControl)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </GlobalMenu>
  <SwipeableEdgeDrawer
    bleedingHeight={56}
    edgeLabel={
      <CloseFormButton
        onClose={[Function]}
      />
    }
    edgeLabelWhenClosed="編集中"
    heightRatio={80}
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    show={false}
  >
    <ArticleForm
      onClickSelectPosition={[Function]}
      park="S"
    />
  </SwipeableEdgeDrawer>
  <GlobalMessage
    closeFormModal={[Function]}
  />
  <ForwardRef(Dialog)
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(DialogTitle)>
      入力内容の変更が保存されていません。
    </ForwardRef(DialogTitle)>
    <ForwardRef(DialogContent)>
      <ForwardRef(DialogContentText)>
        現在編集中の投稿を「下書き保存」、「保存して公開」、または「編集を破棄」してからもう一度お試しください。
      </ForwardRef(DialogContentText)>
    </ForwardRef(DialogContent)>
    <ForwardRef(DialogActions)>
      <ForwardRef(Button)
        onClick={[Function]}
        variant="outlined"
      >
        閉じる
      </ForwardRef(Button)>
    </ForwardRef(DialogActions)>
  </ForwardRef(Dialog)>
</ForwardRef(Box)>
`;

exports[`Shallow Snapshot Tests pc view 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "width": "calc(100% - 500px)",
    }
  }
>
  <GlobalMenu
    topBarPosition="static"
  >
    <ForwardRef(Box)
      sx={
        Object {
          "height": "calc(100vh - 66px)",
          "position": "relative",
        }
      }
    >
      <TriviaMap
        endToSelectPosition={[Function]}
        hiddenMarkerIds={Array []}
        newMarkerMode={false}
        park="S"
        shouldCurrentPositionAsyncWithForm={true}
      />
      <FloatingButton
        color="primary"
        disabled={false}
        icon="add-marker"
        onClick={[Function]}
      />
      <LoadingProgressBar />
      <ForwardRef(Box)
        sx={
          Object {
            "backgroundColor": "white",
            "borderColor": "#9e9e9e",
            "borderRadius": 2,
            "borderStyle": "solid",
            "bottom": "auto",
            "left": "auto",
            "margin": 0,
            "padding": 1,
            "position": "fixed",
            "right": 520,
            "top": 70,
            "zIndex": 1000,
          }
        }
      >
        <ForwardRef(FormControl)
          component="fieldset"
        >
          <ForwardRef(RadioGroup)
            onChange={[Function]}
            value="S"
          >
            <ForwardRef(FormControlLabel)
              control={<ForwardRef(Radio) />}
              label="ランド"
              value="L"
            />
            <ForwardRef(FormControlLabel)
              control={<ForwardRef(Radio) />}
              label="シー"
              value="S"
            />
          </ForwardRef(RadioGroup)>
        </ForwardRef(FormControl)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </GlobalMenu>
  <ForwardRef(Drawer)
    anchor="right"
    open={true}
    sx={
      Object {
        "& .MuiDrawer-paper": Object {
          "width": 500,
        },
        "flexShrink": 0,
        "width": 500,
      }
    }
    variant="persistent"
  >
    <ArticleForm
      onClickSelectPosition={[Function]}
      onClose={[Function]}
      park="S"
    />
  </ForwardRef(Drawer)>
  <GlobalMessage
    closeFormModal={[Function]}
  />
  <ForwardRef(Dialog)
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(DialogTitle)>
      入力内容の変更が保存されていません。
    </ForwardRef(DialogTitle)>
    <ForwardRef(DialogContent)>
      <ForwardRef(DialogContentText)>
        現在編集中の投稿を「下書き保存」、「保存して公開」、または「編集を破棄」してからもう一度お試しください。
      </ForwardRef(DialogContentText)>
    </ForwardRef(DialogContent)>
    <ForwardRef(DialogActions)>
      <ForwardRef(Button)
        onClick={[Function]}
        variant="outlined"
      >
        閉じる
      </ForwardRef(Button)>
    </ForwardRef(DialogActions)>
  </ForwardRef(Dialog)>
</ForwardRef(Box)>
`;
